src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/generate_proposals_ref.cl

```diff
@@ -0,0 +1,318 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#if INPUT0_TYPE_SIZE == 2 //f16
+    #define HALF_ONE 0.5h
+    #define ZERO 0.0h
+
+    #ifdef NORMALIZED
+      #define COORDINATES_OFFSET  0.0h
+    #else
+      #define COORDINATES_OFFSET  1.0h
+    #endif
+
+#else
+
+    #define HALF_ONE 0.5f
+    #define ZERO 0.0f
+
+    #ifdef NORMALIZED
+        #define COORDINATES_OFFSET  0.0f
+    #else
+        #define COORDINATES_OFFSET  1.0f
+    #endif
+
+#endif
+
+#define ROI_SIZE 4
+
+#ifdef GENERATE_PROPOSALS_STAGE_0
+
+// 0. Refine anchors
+KERNEL(generate_proposals_ref_stage_0)
+(const __global INPUT0_TYPE* im_info,
+ const __global INPUT0_TYPE* anchors,
+ const __global INPUT0_TYPE* deltas,
+ const __global INPUT0_TYPE* scores,
+ __global INPUT0_TYPE* proposals) {
```

Comment:
I think dedicated types should be used for each tensor here and below, i.e. INPUT1_TYPE, INPUT2_TYPE, OUTPUT_TYPE